htlcswitch/interceptable_switch.go

```diff
@@ -35,75 +35,195 @@ type InterceptableSwitch struct {
+		resolutionChan:          make(chan *FwdResolution),
+
+		quit: make(chan struct{}),
+	}
 }
 
 // SetInterceptor sets the ForwardInterceptor to be used.
 func (s *InterceptableSwitch) SetInterceptor(
 	interceptor ForwardInterceptor) {
 
-	s.Lock()
-	defer s.Unlock()
-	s.fwdInterceptor = interceptor
+	select {
+	case s.interceptorRegistration <- interceptor:
+
+	case <-s.quit:
+	}
+}
+
+func (s *InterceptableSwitch) Run() error {
+	for {
+		// If an interceptor is required and not set, block.
+		if s.requireInterceptor {
+			for s.interceptor == nil {
+				select {
+				case interceptor := <-s.interceptorRegistration:
+					s.interceptor = interceptor
+
+				case <-s.quit:
+					return nil
+				}
+			}
+		}
+
+		select {
+		case interceptor := <-s.interceptorRegistration:
+			s.interceptor = interceptor
+
+			// If a new interceptor is registered or an interceptor
+			// is always required, don't release held forwards.
+			if interceptor != nil || s.requireInterceptor {
+				break
+			}
+
+			// Release held forwards.
+			log.Infof("RPC interceptor disconnected, resolving held packets")
+			for _, fwd := range s.holdForwards {
+				if err := fwd.resume(); err != nil {
+					log.Errorf("failed to resume hold forward %v", err)
+				}
+			}
+			s.holdForwards = make(map[channeldb.CircuitKey]*interceptedForward)
+
+		case packets := <-s.intercepted:
+			var notIntercepted []*htlcPacket
+			for _, p := range packets {
+				if s.interceptor == nil ||
+					!s.interceptForward(p) {
+
+					notIntercepted = append(notIntercepted, p)
+				}
+			}
+			return s.htlcSwitch.ForwardPackets(
+				nil, notIntercepted...,
+			)
+
+		case res := <-s.resolutionChan:
+			intercepted, ok := s.holdForwards[res.Key]
+			if !ok {
+				break
+			}
+			delete(s.holdForwards, res.Key)
+
+			switch res.Action {
+			case FwdActionResume:
+				intercepted.resume()
+
+			case FwdActionSettle:
+				intercepted.settle(res.Preimage)
+
+			case FwdActionFail:
+				if len(res.FailureMessage) > 0 {
+					intercepted.fail(res.FailureMessage)
+				} else {
+					intercepted.failWithCode(res.FailureCode)
+				}
+			}
+
+		case <-s.quit:
+			return nil
+		}
+	}
+}
+
+type FwdAction int
+
+const (
+	FwdActionResume FwdAction = iota
+	FwdActionSettle
+	FwdActionFail
+)
+
+type FwdResolution struct {
+	Key channeldb.CircuitKey
+
+	Action         FwdAction
+	Preimage       lntypes.Preimage
+	FailureMessage []byte
+	FailureCode    lnwire.FailCode
+}
+
+func (s *InterceptableSwitch) Resolve(res *FwdResolution) {
+	select {
+	case s.resolutionChan <- res:
+	case <-s.quit:
+	}
 }
 
 // ForwardPackets attempts to forward the batch of htlcs through the
 // switch, any failed packets will be returned to the provided
 // ChannelLink. The link's quit signal should be provided to allow
 // cancellation of forwarding during link shutdown.
-func (s *InterceptableSwitch) ForwardPackets(linkQuit chan struct{},
+func (s *InterceptableSwitch) ForwardPackets(_ chan struct{},
```

Comment:
Upon further inspection, it seems that `linkQuit` is not worth the hassle to thread through.

Reply:
This effectively removes the feature entirely, doesn't it? Since packets always come through here. I'm not sure about the merits `linkQuit` has, or lacks, but it IMO this is not the PR to kill it or the way to do it. If desired, it should be removed completely from htlcSwitch rather than silently disused.

Reply:
Added commit deleting `linkQuit`.

Reply:
I also deleted a complicated test case `TestChannelLinkShutdownDuringForward` testing `linkQuit`. To me it seems that the case where the switch is completely blocked isn't realistic, but maybe someone with more context can weigh in.